// @azure/msal-browser/dist/index.mjs
/*! @azure/msal-browser v5.6.3 2026-04-01 */
'use strict';
import * as BrowserUtils from './utils/BrowserUtils.mjs';
export { BrowserUtils };
import { Constants } from '@azure/msal-common/browser';
export { AuthError, AuthErrorCodes, AuthenticationHeaderParser, AzureCloudInstance, ClientAuthError, ClientAuthErrorCodes, ClientConfigurationError, ClientConfigurationErrorCodes, InteractionRequiredAuthError, InteractionRequiredAuthErrorCodes, LogLevel, Logger, ProtocolMode, ServerError, StubPerformanceClient, enforceResourceParameter } from '@azure/msal-common/browser';
export { PublicClientApplication, createNestablePublicClientApplication, createStandardPublicClientApplication } from './app/PublicClientApplication.mjs';
export { DEFAULT_IFRAME_TIMEOUT_MS } from './config/Configuration.mjs';
export { ApiId, BrowserCacheLocation, CacheLookupPolicy, InteractionStatus, InteractionType, WrapperSKU } from './utils/BrowserConstants.mjs';
export { BrowserAuthError } from './error/BrowserAuthError.mjs';
export { BrowserConfigurationAuthError } from './error/BrowserConfigurationAuthError.mjs';
export { stubbedPublicClientApplication } from './app/IPublicClientApplication.mjs';
export { NavigationClient } from './navigation/NavigationClient.mjs';
export { loadExternalTokens } from './cache/TokenCache.mjs';
export { MemoryStorage } from './cache/MemoryStorage.mjs';
export { LocalStorage } from './cache/LocalStorage.mjs';
export { SessionStorage } from './cache/SessionStorage.mjs';
export { EventMessageUtils } from './event/EventMessage.mjs';
export { EventType } from './event/EventType.mjs';
export { EventHandler } from './event/EventHandler.mjs';
export { SignedHttpRequest } from './crypto/SignedHttpRequest.mjs';
export { BrowserPerformanceClient } from './telemetry/BrowserPerformanceClient.mjs';
export { BrowserPerformanceMeasurement } from './telemetry/BrowserPerformanceMeasurement.mjs';
import * as BrowserRootPerformanceEvents from './telemetry/BrowserRootPerformanceEvents.mjs';
export { BrowserRootPerformanceEvents };
export { version } from './packageMetadata.mjs';
export { isPlatformBrokerAvailable } from './broker/nativeBroker/PlatformAuthProvider.mjs';
import * as BrowserAuthErrorCodes from './error/BrowserAuthErrorCodes.mjs';
export { BrowserAuthErrorCodes };
import * as BrowserConfigurationAuthErrorCodes from './error/BrowserConfigurationAuthErrorCodes.mjs';
export { BrowserConfigurationAuthErrorCodes };

/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
/**
 * @packageDocumentation
 * @module @azure/msal-browser
 */
// Common constants
const AuthenticationScheme = Constants.AuthenticationScheme;
const ResponseMode = Constants.ResponseMode;
const PromptValue = Constants.PromptValue;
const JsonWebTokenTypes = Constants.JsonWebTokenTypes;
const OIDC_DEFAULT_SCOPES = Constants.OIDC_DEFAULT_SCOPES;

export { AuthenticationScheme, JsonWebTokenTypes, OIDC_DEFAULT_SCOPES, PromptValue, ResponseMode };
