// @azure/msal-common/dist-browser/index.mjs
/*! @azure/msal-common v16.4.1 2026-04-01 */
'use strict';
export { AuthorizationCodeClient } from './client/AuthorizationCodeClient.mjs';
export { RefreshTokenClient } from './client/RefreshTokenClient.mjs';
export { SilentFlowClient } from './client/SilentFlowClient.mjs';
export { DEFAULT_SYSTEM_OPTIONS, buildClientConfiguration } from './config/ClientConfiguration.mjs';
export { buildTenantProfile, tenantIdMatchesHomeTenant, updateAccountTenantProfileData } from './account/AccountInfo.mjs';
export { getTenantIdFromIdTokenClaims } from './account/TokenClaims.mjs';
export { CcsCredentialType } from './account/CcsCredential.mjs';
export { buildClientInfo, buildClientInfoFromHomeAccountId } from './account/ClientInfo.mjs';
export { Authority, buildStaticAuthorityOptions, formatAuthorityUri } from './authority/Authority.mjs';
export { AzureCloudInstance } from './authority/AuthorityOptions.mjs';
export { AuthorityType } from './authority/AuthorityType.mjs';
export { ProtocolMode } from './authority/ProtocolMode.mjs';
export { CacheManager, DefaultStorageClass } from './cache/CacheManager.mjs';
export { StubbedNetworkModule } from './network/INetworkModule.mjs';
export { ThrottlingUtils } from './network/ThrottlingUtils.mjs';
export { getRequestThumbprint } from './network/RequestThumbprint.mjs';
export { UrlString } from './url/UrlString.mjs';
export { DEFAULT_CRYPTO_IMPLEMENTATION } from './crypto/ICrypto.mjs';
import * as Authorize from './protocol/Authorize.mjs';
export { Authorize as AuthorizeProtocol };
import * as Token from './protocol/Token.mjs';
export { Token as TokenProtocol };
export { enforceResourceParameter } from './request/BaseAuthRequest.mjs';
import * as RequestParameterBuilder from './request/RequestParameterBuilder.mjs';
export { RequestParameterBuilder };
export { ResponseHandler, buildAccountToCache } from './response/ResponseHandler.mjs';
export { ScopeSet } from './request/ScopeSet.mjs';
export { AuthenticationHeaderParser } from './request/AuthenticationHeaderParser.mjs';
export { LogLevel, Logger } from './logger/Logger.mjs';
export { InteractionRequiredAuthError, createInteractionRequiredAuthError } from './error/InteractionRequiredAuthError.mjs';
import * as InteractionRequiredAuthErrorCodes from './error/InteractionRequiredAuthErrorCodes.mjs';
export { InteractionRequiredAuthErrorCodes };
export { AuthError, createAuthError } from './error/AuthError.mjs';
import * as AuthErrorCodes from './error/AuthErrorCodes.mjs';
export { AuthErrorCodes };
export { PlatformBrokerError } from './error/PlatformBrokerError.mjs';
export { ServerError } from './error/ServerError.mjs';
export { NetworkError, createNetworkError } from './error/NetworkError.mjs';
export { CacheError, createCacheError } from './error/CacheError.mjs';
import * as CacheErrorCodes from './error/CacheErrorCodes.mjs';
export { CacheErrorCodes };
export { ClientAuthError, createClientAuthError } from './error/ClientAuthError.mjs';
import * as ClientAuthErrorCodes from './error/ClientAuthErrorCodes.mjs';
export { ClientAuthErrorCodes };
export { ClientConfigurationError, createClientConfigurationError } from './error/ClientConfigurationError.mjs';
import * as ClientConfigurationErrorCodes from './error/ClientConfigurationErrorCodes.mjs';
export { ClientConfigurationErrorCodes };
import * as Constants from './utils/Constants.mjs';
export { Constants };
export { StringUtils } from './utils/StringUtils.mjs';
import * as ProtocolUtils from './utils/ProtocolUtils.mjs';
export { ProtocolUtils };
export { ServerTelemetryManager } from './telemetry/server/ServerTelemetryManager.mjs';
export { version } from './packageMetadata.mjs';
export { invoke, invokeAsync } from './utils/FunctionWrappers.mjs';
import * as AuthToken from './account/AuthToken.mjs';
export { AuthToken };
import * as AuthorityFactory from './authority/AuthorityFactory.mjs';
export { AuthorityFactory };
import * as CacheHelpers from './cache/utils/CacheHelpers.mjs';
export { CacheHelpers };
import * as TimeUtils from './utils/TimeUtils.mjs';
export { TimeUtils };
import * as UrlUtils from './utils/UrlUtils.mjs';
export { UrlUtils };
import * as AADServerParamKeys from './constants/AADServerParamKeys.mjs';
export { AADServerParamKeys };
import * as AccountEntityUtils from './cache/utils/AccountEntityUtils.mjs';
export { AccountEntityUtils };
export { JoseHeader } from './crypto/JoseHeader.mjs';
export { IntFields, PerformanceEventStatus } from './telemetry/performance/PerformanceEvent.mjs';
import * as PerformanceEvents from './telemetry/performance/PerformanceEvents.mjs';
export { PerformanceEvents };
export { PerformanceClient } from './telemetry/performance/PerformanceClient.mjs';
export { StubPerformanceClient } from './telemetry/performance/StubPerformanceClient.mjs';
export { PopTokenGenerator } from './crypto/PopTokenGenerator.mjs';
export { TokenCacheContext } from './cache/persistence/TokenCacheContext.mjs';
import * as ClientAssertionUtils from './utils/ClientAssertionUtils.mjs';
export { ClientAssertionUtils };
export { getClientAssertion } from './utils/ClientAssertionUtils.mjs';
