// @azure/arm-storage/dist/esm/operations/objectReplicationPoliciesOperations.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers.js";
import * as Parameters from "../models/parameters.js";
/// <reference lib="esnext.asynciterable" />
/** Class containing ObjectReplicationPoliciesOperations operations. */
export class ObjectReplicationPoliciesOperationsImpl {
    client;
    /**
     * Initialize a new instance of the class ObjectReplicationPoliciesOperations class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * List the object replication policies associated with the storage account.
     * @param resourceGroupName The name of the resource group within the user's subscription. The name is
     *                          case insensitive.
     * @param accountName The name of the storage account within the specified resource group. Storage
     *                    account names must be between 3 and 24 characters in length and use numbers and lower-case letters
     *                    only.
     * @param options The options parameters.
     */
    list(resourceGroupName, accountName, options) {
        const iter = this.listPagingAll(resourceGroupName, accountName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, accountName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, accountName, options, _settings) {
        let result;
        result = await this._list(resourceGroupName, accountName, options);
        yield result.value || [];
    }
    async *listPagingAll(resourceGroupName, accountName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, accountName, options)) {
            yield* page;
        }
    }
    /**
     * List the object replication policies associated with the storage account.
     * @param resourceGroupName The name of the resource group within the user's subscription. The name is
     *                          case insensitive.
     * @param accountName The name of the storage account within the specified resource group. Storage
     *                    account names must be between 3 and 24 characters in length and use numbers and lower-case letters
     *                    only.
     * @param options The options parameters.
     */
    _list(resourceGroupName, accountName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, accountName, options }, listOperationSpec);
    }
    /**
     * Get the object replication policy of the storage account by policy ID.
     * @param resourceGroupName The name of the resource group within the user's subscription. The name is
     *                          case insensitive.
     * @param accountName The name of the storage account within the specified resource group. Storage
     *                    account names must be between 3 and 24 characters in length and use numbers and lower-case letters
     *                    only.
     * @param objectReplicationPolicyId For the destination account, provide the value 'default'. Configure
     *                                  the policy on the destination account first. For the source account, provide the value of the policy
     *                                  ID that is returned when you download the policy that was defined on the destination account. The
     *                                  policy is downloaded as a JSON file.
     * @param options The options parameters.
     */
    get(resourceGroupName, accountName, objectReplicationPolicyId, options) {
        return this.client.sendOperationRequest({ resourceGroupName, accountName, objectReplicationPolicyId, options }, getOperationSpec);
    }
    /**
     * Create or update the object replication policy of the storage account.
     * @param resourceGroupName The name of the resource group within the user's subscription. The name is
     *                          case insensitive.
     * @param accountName The name of the storage account within the specified resource group. Storage
     *                    account names must be between 3 and 24 characters in length and use numbers and lower-case letters
     *                    only.
     * @param objectReplicationPolicyId For the destination account, provide the value 'default'. Configure
     *                                  the policy on the destination account first. For the source account, provide the value of the policy
     *                                  ID that is returned when you download the policy that was defined on the destination account. The
     *                                  policy is downloaded as a JSON file.
     * @param properties The object replication policy set to a storage account. A unique policy ID will be
     *                   created if absent.
     * @param options The options parameters.
     */
    createOrUpdate(resourceGroupName, accountName, objectReplicationPolicyId, properties, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            accountName,
            objectReplicationPolicyId,
            properties,
            options,
        }, createOrUpdateOperationSpec);
    }
    /**
     * Deletes the object replication policy associated with the specified storage account.
     * @param resourceGroupName The name of the resource group within the user's subscription. The name is
     *                          case insensitive.
     * @param accountName The name of the storage account within the specified resource group. Storage
     *                    account names must be between 3 and 24 characters in length and use numbers and lower-case letters
     *                    only.
     * @param objectReplicationPolicyId For the destination account, provide the value 'default'. Configure
     *                                  the policy on the destination account first. For the source account, provide the value of the policy
     *                                  ID that is returned when you download the policy that was defined on the destination account. The
     *                                  policy is downloaded as a JSON file.
     * @param options The options parameters.
     */
    delete(resourceGroupName, accountName, objectReplicationPolicyId, options) {
        return this.client.sendOperationRequest({ resourceGroupName, accountName, objectReplicationPolicyId, options }, deleteOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Storage/storageAccounts/{accountName}/objectReplicationPolicies",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ObjectReplicationPolicies,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Storage/storageAccounts/{accountName}/objectReplicationPolicies/{objectReplicationPolicyId}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ObjectReplicationPolicy,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId,
        Parameters.objectReplicationPolicyId,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Storage/storageAccounts/{accountName}/objectReplicationPolicies/{objectReplicationPolicyId}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.ObjectReplicationPolicy,
        },
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    requestBody: Parameters.properties3,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId,
        Parameters.objectReplicationPolicyId,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Storage/storageAccounts/{accountName}/objectReplicationPolicies/{objectReplicationPolicyId}",
    httpMethod: "DELETE",
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.ErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.accountName,
        Parameters.subscriptionId,
        Parameters.objectReplicationPolicyId,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
