// @azure/msal-node/dist/index.mjs
/*! @azure/msal-node v5.1.2 2026-04-01 */
'use strict';
import * as internals from './internals.mjs';
export { internals };
import { Constants } from '@azure/msal-common/node';
export { AuthError, AuthErrorCodes, AzureCloudInstance, ClientAuthError, ClientAuthErrorCodes, ClientConfigurationError, ClientConfigurationErrorCodes, InteractionRequiredAuthError, InteractionRequiredAuthErrorCodes, LogLevel, Logger, ProtocolMode, ServerError, TokenCacheContext } from '@azure/msal-common/node';
export { PublicClientApplication } from './client/PublicClientApplication.mjs';
export { ConfidentialClientApplication } from './client/ConfidentialClientApplication.mjs';
export { ManagedIdentityApplication } from './client/ManagedIdentityApplication.mjs';
export { ClientAssertion } from './client/ClientAssertion.mjs';
export { TokenCache } from './cache/TokenCache.mjs';
export { DistributedCachePlugin } from './cache/distributed/DistributedCachePlugin.mjs';
export { ManagedIdentitySourceNames } from './utils/Constants.mjs';
export { CryptoProvider } from './crypto/CryptoProvider.mjs';
export { version } from './packageMetadata.mjs';

/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
/**
 * @packageDocumentation
 * @module @azure/msal-node
 */
/**
 * Warning: This set of exports is purely intended to be used by other MSAL libraries, and should be considered potentially unstable. We strongly discourage using them directly, you do so at your own risk.
 * Breaking changes to these APIs will be shipped under a minor version, instead of a major version.
 */
const PromptValue = Constants.PromptValue;
const ResponseMode = Constants.ResponseMode;

export { PromptValue, ResponseMode };
